topology/intel/dmic/dmic-nhlt.c

```diff
@@ -0,0 +1,1827 @@
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+  The full GNU General Public License is included in this distribution
+  in the file called LICENSE.GPL.
+*/
+
+#include <stdint.h>
+#include <errno.h>
+#include <stdio.h>
+#include <alsa/input.h>
+#include <alsa/output.h>
+#include <alsa/conf.h>
+#include <alsa/error.h>
+#include "../nhlt-internal.h"
+#include "dmic-nhlt.h"
+#include "pdm_decim_table.h"
+#include "dmic_macros.h"
+
+/* structs for dmic internal calculations */
+struct dmic_calc_decim_modes {
+        int16_t clkdiv[DMIC_MAX_MODES];
+        int16_t mcic[DMIC_MAX_MODES];
+        int16_t mfir[DMIC_MAX_MODES];
+        int num_of_modes;
+};
+
+struct dmic_calc_matched_modes {
+        int16_t clkdiv[DMIC_MAX_MODES];
+        int16_t mcic[DMIC_MAX_MODES];
+        int16_t mfir_a[DMIC_MAX_MODES];
+        int16_t mfir_b[DMIC_MAX_MODES];
+        int num_of_modes;
+};
+
+struct dmic_calc_configuration {
+        struct pdm_decim *fir_a;
+        struct pdm_decim *fir_b;
+        int clkdiv;
+        int mcic;
+        int mfir_a;
+        int mfir_b;
+        int cic_shift;
+        int fir_a_shift;
+        int fir_b_shift;
+        int fir_a_length;
+        int fir_b_length;
+        int32_t fir_a_scale;
+        int32_t fir_b_scale;
+};
+
+#ifdef DMIC_NHLT_DEBUG
+
+#define MAX_NHLT_BUFFER_SIZE 25000
+
+/* print blob as bytes hex string like: 0x11,0xff,0xff,0xff etc. */
+static void dmic_print_bytes_as_hex(uint8_t *src, size_t size)
+{
+	tplg_pp_debug("printing dmic vendor blob as bytes:");
+	char nhlt_string_buffer[MAX_NHLT_BUFFER_SIZE];
+	char *nhlt_p = &nhlt_string_buffer[0];
+	int i;
+
+	for (i = 0; i < size; i++) {
+		snprintf(nhlt_p, 6, "0x%02x,", *src);
+		nhlt_p += 5;
+		src++;
+	}
+
+	/* remove the last comma... */
+	nhlt_p--;
+	*nhlt_p = '\0';
+
+	printf("%s\n", nhlt_string_buffer);
+}
+
+/* print blob as 32 bit integer hex string like: 0xffffffff,0x00000010 etc. */
+static void dmic_print_integers_as_hex(uint32_t *src, size_t size)
+{
+	tplg_pp_debug("printing dmic vendor blob as integers:");
+
+	char nhlt_string_buffer[MAX_NHLT_BUFFER_SIZE];
+	char *nhlt_p = &nhlt_string_buffer[0];
+	int i;
+
+	for (i = 0; i < size; i++) {
+		snprintf(nhlt_p, 12, "0x%08x,", *src);
+		nhlt_p += 11;
+		src++;
+	}
+
+	/* remove the last comma... */
+	nhlt_p--;
+	*nhlt_p = '\0';
+
+	printf("%s\n", nhlt_string_buffer);
+
+	tplg_pp_debug("");
+}
+
+static void dmic_print_internal(struct intel_nhlt_params *nhlt)
+{
+	int i, j, line, lines, remind;
+
+	tplg_pp_debug("printing dmic nhlt internal data");
+
+	/* top level struct */
+	tplg_pp_debug("gateway attributes: %u", nhlt->dmic_blob.gateway_attributes);
```

Comment:
Nit, "0x%08x" would make a nicer format for debug output. Many of the bit fields are aligned at hex digit boundaries so hex is easier to decode by eye.

Reply:
ok